syntax = "proto3";

package commu;

message TaskReq {
    int32 task_id = 1;
    string member_id = 2;
}

message MetadataResp {
    string name = 1;
    string type = 2;
    int32 secure_level = 3;
    string algorithm = 4;
    repeated string members = 5;
}

message JoinResp {
    bool success = 1;
}

message RoundResp {
    int32 round_id = 1;
}

message FileReq {
    string type = 1;
    int32 task_id = 2;
    string member_id = 3;
    int32 round_id = 4;
}

message FileResp {
    string filename = 1;
    bytes content = 2;
}

message ResultReq {
    int32 task_id = 1;
    string member_id = 2;
    int32 round_id = 3;
    string type = 4;
    bytes content = 5;
}

message ResultResp {
    string type = 1;
    bytes content = 2;
}

service Commu {
    rpc GetMetadata(TaskReq) returns (MetadataResp) {}
    rpc JoinTask(TaskReq) returns (JoinResp) {}
    rpc GetRound(TaskReq) returns (RoundResp) {}
    rpc GetFile(FileReq) returns (stream FileResp) {}
    rpc UploadResult(stream ResultReq) returns (stream ResultResp) {} 
}